File: src/frontend/Encryption/EncryptionTest.java
package src.frontend.Encryption;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.ValueSource;

import javax.crypto.AEADBadTagException;
import javax.crypto.SecretKey;
import java.security.SecureRandom;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;
import static src.frontend.Encryption.Encryption.*;

class EncryptionTest {

    private byte[] salt;
    private String password;
    private SecretKey key;

    /**
     * Sets up the test environment before each test.
     */
    @BeforeEach
    public void setUp() throws Exception {
        salt = generateSalt();
        password = "password";
        key = generateKeyFromPassword(password, salt);
    }

    @Nested
    @DisplayName("Encryption and Decryption Tests")
    class EncryptionDecryptionTests {

        static Stream<Arguments> dataProviderForEncryption() {

            return Stream.of(
                    Arguments.of("Short text"),
                    Arguments.of(""),
                    Arguments.of("A longer piece of text to test the encryption and decryption functionality."),
                    Arguments.of(generateRandomString(1024)), // 1 KB
                    Arguments.of(generateRandomString(10 * 1024 * 1024)) // 10 MB
            );
        }

        private static String generateRandomString(int length) {

            SecureRandom random = new SecureRandom();
            StringBuilder sb = new StringBuilder(length);
            for (int i = 0; i < length; i++) {
                int randomAscii = 33 + random.nextInt(94);
                sb.append((char) randomAscii);
            }
            return sb.toString();
        }

        @ParameterizedTest(name = "Data set: \"{0}\"")
        @MethodSource("dataProviderForEncryption")
        @DisplayName("Test encryption and decryption with multiple data sets")
        void testEncryptionAndDecryptionWithMultipleDataSets(String data) throws Exception {

            // setup
            byte[] dataBytes = data.getBytes();

            // act
            Pair<byte[], byte[]> encryptedData = encrypt(dataBytes, key);
            byte[] decryptedData = decrypt(key, encryptedData.first(), encryptedData.second());

            // assert
            assertArrayEquals(dataBytes, decryptedData, "Decrypted data should match the original");
        }

        @Test
        @DisplayName("Encrypting the same data multiple times produces different ciphertexts")
        void testIVRandomness() throws Exception {

            // setup
            String data = "my secret";

            // act
            Pair<byte[], byte[]> encryptedData1 = encrypt(data.getBytes(), key);
            Pair<byte[], byte[]> encryptedData2 = encrypt(data.getBytes(), key);

            // assert
            assertFalse(
                    java.util.Arrays.equals(encryptedData1.first(), encryptedData2.first()),
                    "IVs should be unique for each encryption"
            );
            assertNotEquals(
                    new String(encryptedData1.second()),
                    new String(encryptedData2.second()),
                    "Ciphertexts should differ due to different IVs"
            );
        }
    }

    @Nested
    @DisplayName("Key Generation Tests")
    class KeyGenerationTests {

        @Test
        @DisplayName("Test key generation consistency with same password and salt")
        void testKeyGenerationConsistency() throws Exception {

            // act
            SecretKey newKey = generateKeyFromPassword(password, salt);

            // assert
            assertArrayEquals(
                    key.getEncoded(), newKey.getEncoded(),
                    "Keys generated with the same password and salt should be identical");
        }

        @Test
        @DisplayName("Test key generation with different salts produces different keys")
        void testKeyGenerationDifferentSalts() throws Exception {

            // setup
            byte[] newSalt = generateSalt();

            // act
            SecretKey newKey = generateKeyFromPassword(password, newSalt);

            // assert
            assertFalse(
                    java.util.Arrays.equals(key.getEncoded(), newKey.getEncoded()),
                    "Keys generated with the same password but different salts should differ"
            );
        }

        @Test
        @DisplayName("Test key generation with different passwords produces different keys")
        void testKeyGenerationDifferentPasswords() throws Exception {

            // setup
            String newPassword = "newPassword";

            // act
            SecretKey newKey = generateKeyFromPassword(newPassword, salt);

            // assert
            assertFalse(
                    java.util.Arrays.equals(key.getEncoded(), newKey.getEncoded()),
                    "Keys generated with different passwords should differ"
            );
        }
    }

    @Nested
    @DisplayName("Salt Generation Tests")
    class SaltGenerationTests {

        @Test
        @DisplayName("Test salt is of correct default length")
        void testDefaultSaltLength() {

            // assert
            assertEquals(10, salt.length, "Default salt length should be 10 bytes");
        }

        @ParameterizedTest(name = "Salt length: {0} bytes")
        @ValueSource(ints = {8, 16, 32})
        @DisplayName("Test salt generation with specified lengths")
        void testSaltGenerationWithSpecifiedLengths(int length) {

            // act
            byte[] generatedSalt = generateSalt(length);

            // assert
            assertEquals(length, generatedSalt.length, "Salt length should match the specified length");
        }

        @Test
        @DisplayName("Test salt randomness")
        void testSaltRandomness() {

            // setup
            int numSalts = 1000;
            Set<String> salts = new HashSet<>();

            // act
            for (int i = 0; i < numSalts; i++) {
                byte[] salt = generateSalt();
                salts.add(java.util.Arrays.toString(salt));
            }

            // assert
            assertEquals(numSalts, salts.size(), "All generated salts should be unique");
        }
    }

    @Nested
    @DisplayName("Exception Handling Tests")
    class ExceptionHandlingTests {

        /**
         * Provides null parameters for encryption tests.
         */
        static Stream<Arguments> nullParametersProviderForEncryption() {

            return Stream.of(
                    Arguments.of(null, "validKey"),
                    Arguments.of("validData", null),
                    Arguments.of(null, null)
            ).map(args -> {
                try {
                    SecretKey key = generateKeyFromPassword("password", generateSalt());
                    byte[] data = "validData".equals(args.get()[0]) ? "Short text".getBytes() : null;
                    SecretKey keyParam = "validKey".equals(args.get()[1]) ? key : null;
                    return Arguments.of(data, keyParam);
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            });
        }

        /**
         * Provides null parameters for decryption tests.
         */
        static Stream<Arguments> nullParametersProviderForDecryption() {

            return Stream.of(
                    Arguments.of(null, "validIV", "validEncryptedData"),
                    Arguments.of("validKey", null, "validEncryptedData"),
                    Arguments.of("validKey", "validIV", null),
                    Arguments.of(null, null, null)
            ).map(args -> {
                try {
                    SecretKey key = generateKeyFromPassword("password", generateSalt());
                    Pair<byte[], byte[]> encryptedData = encrypt("Short text".getBytes(), key);
                    SecretKey keyParam = "validKey".equals(args.get()[0]) ? key : null;
                    byte[] iv = "validIV".equals(args.get()[1]) ? encryptedData.first() : null;
                    byte[] cipherText = "validEncryptedData".equals(args.get()[2]) ? encryptedData.second() : null;
                    return Arguments.of(keyParam, iv, cipherText);
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            });
        }

        /**
         * Provides null parameters for key generation tests.
         */
        static Stream<Arguments> nullParametersProviderForKeyGeneration() {

            return Stream.of(
                    Arguments.of(null, "validSalt"),
                    Arguments.of("validPassword", null),
                    Arguments.of(null, null)
            ).map(args -> {
                String passwordParam = "validPassword".equals(args.get()[0]) ? "password" : null;
                byte[] saltParam = "validSalt".equals(args.get()[1]) ? generateSalt() : null;
                return Arguments.of(passwordParam, saltParam);
            });
        }

        @Test
        @DisplayName("Test decryption with invalid key throws AEADBadTagException")
        void testDecryptionWithInvalidKey() throws Exception {

            // setup
            SecretKey invalidKey = generateKeyFromPassword("invalidPassword", salt);
            String data = "Short text.";

            // act
            Pair<byte[], byte[]> encryptedData = encrypt(data.getBytes(), key);

            // assert
            AEADBadTagException exception = assertThrows(
                    AEADBadTagException.class,
                    () -> decrypt(invalidKey, encryptedData.first(), encryptedData.second()),
                    "Decryption with an invalid key should throw AEADBadTagException"
            );
            assertEquals("Tag mismatch", exception.getMessage(), "Exception message should indicate tag mismatch");
        }

        @Test
        @DisplayName("Test decryption with invalid IV throws AEADBadTagException")
        void testDecryptionWithInvalidIV() throws Exception {

            // setup
            String data = "Short text";
            Pair<byte[], byte[]> encryptedData = encrypt(data.getBytes(), key);
            encryptedData.first()[0] ^= (byte) 0xFF;

            // assert
            AEADBadTagException exception = assertThrows(
                    AEADBadTagException.class,
                    () -> decrypt(key, encryptedData.first(), encryptedData.second()),
                    "Decryption with an invalid IV should throw AEADBadTagException"
            );
            assertEquals("Tag mismatch", exception.getMessage(), "Exception message should indicate tag mismatch");
        }

        @Test
        @DisplayName("Test decryption with corrupted ciphertext throws AEADBadTagException")
        void testDecryptionWithCorruptedCiphertext() throws Exception {

            // setup
            String data = "Short text";
            Pair<byte[], byte[]> encryptedData = encrypt(data.getBytes(), key);
            encryptedData.second()[0] ^= (byte) 0xFF;

            // assert
            AEADBadTagException exception = assertThrows(
                    AEADBadTagException.class,
                    () -> decrypt(key, encryptedData.first(), encryptedData.second()),
                    "Decryption with corrupted ciphertext should throw AEADBadTagException"
            );
            assertEquals("Tag mismatch", exception.getMessage(), "Exception message should indicate tag mismatch");
        }

        @ParameterizedTest(name = "Null parameter: {0}")
        @MethodSource("nullParametersProviderForEncryption")
        @DisplayName("Test encryption with null parameters throws NullPointerException")
        void testEncryptionWithNullParameters(Object param1, Object param2) {

            // setup
            byte[] data = param1 instanceof byte[] ? (byte[]) param1 : null;
            SecretKey key = param2 instanceof SecretKey ? (SecretKey) param2 : null;

            // assert
            assertThrows(
                    NullPointerException.class,
                    () -> encrypt(data, key),
                    "Encryption with null parameters should throw NullPointerException");
        }

        @ParameterizedTest(name = "Null parameter: {0}")
        @MethodSource("nullParametersProviderForDecryption")
        @DisplayName("Test decryption with null parameters throws NullPointerException")
        void testDecryptionWithNullParameters(Object param1, Object param2, Object param3) {

            // setup
            SecretKey key = param1 instanceof SecretKey ? (SecretKey) param1 : null;
            byte[] iv = param2 instanceof byte[] ? (byte[]) param2 : null;
            byte[] cipherText = param3 instanceof byte[] ? (byte[]) param3 : null;

            // assert
            assertThrows(
                    NullPointerException.class,
                    () -> decrypt(key, iv, cipherText),
                    "Decryption with null parameters should throw NullPointerException"
            );
        }

        @ParameterizedTest(name = "Null parameter: {0}")
        @MethodSource("nullParametersProviderForKeyGeneration")
        @DisplayName("Test key generation with null parameters throws NullPointerException")
        void testKeyGenerationWithNullParameters(Object param1, Object param2) {

            // setup
            String password = param1 instanceof String ? (String) param1 : null;
            byte[] salt = param2 instanceof byte[] ? (byte[]) param2 : null;

            // assert
            assertThrows(
                    NullPointerException.class,
                    () -> generateKeyFromPassword(password, salt),
                    "Encryption with null parameters should throw NullPointerException"
            );
        }
    }
}

File: src/tests/EncryptionTest.kt
import org.junit.jupiter.api.*
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.junit.jupiter.params.provider.ValueSource
import expo.modules.encryption.Encryption.decrypt
import expo.modules.encryption.Encryption.encrypt
import expo.modules.encryption.Encryption.generateKeyFromPassword
import expo.modules.encryption.Encryption.generateSalt
import java.security.SecureRandom
import java.util.stream.Stream
import javax.crypto.AEADBadTagException
import javax.crypto.SecretKey

internal class EncryptionTest {
    private lateinit var salt: ByteArray
    private lateinit var password: String
    private lateinit var key: SecretKey

    /**
     * Sets up the test environment before each test.
     */
    @BeforeEach
    @Throws(Exception::class)
    fun setUp() {
        salt = generateSalt()
        password = "password"
        key = generateKeyFromPassword(password, salt)
    }

    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
    @Nested
    @DisplayName("Encryption and Decryption Tests")
    internal inner class EncryptionDecryptionTests {
        @ParameterizedTest(name = "Data set: \"{0}\"")
        @MethodSource("dataProviderForEncryption")
        @DisplayName("Test encryption and decryption with multiple data sets")
        @Throws(
            Exception::class
        )
        fun testEncryptionAndDecryptionWithMultipleDataSets(data: String) {
            // setup

            val dataBytes = data.toByteArray()

            // act
            val encryptedData = encrypt(dataBytes, key)
            val decryptedData = decrypt(key, encryptedData.first, encryptedData.second)

            // assert
            Assertions.assertArrayEquals(dataBytes, decryptedData, "Decrypted data should match the original")
        }

        @Test
        @DisplayName("Encrypting the same data multiple times produces different ciphertexts")
        @Throws(
            Exception::class
        )
        fun testIVRandomness() {
            // setup

            val data = "my secret"

            // act
            val encryptedData1 = encrypt(data.toByteArray(), key)
            val encryptedData2 = encrypt(data.toByteArray(), key)

            // assert
            Assertions.assertFalse(
                encryptedData1.first.contentEquals(encryptedData2.first),
                "IVs should be unique for each encryption"
            )
            Assertions.assertNotEquals(
                String(encryptedData1.second),
                String(encryptedData2.second),
                "Ciphertexts should differ due to different IVs"
            )
        }

        fun dataProviderForEncryption(): Stream<Arguments> {
            return Stream.of(
                Arguments.of("Short text"),
                Arguments.of(""),
                Arguments.of("A longer piece of text to test the encryption and decryption functionality."),
                Arguments.of(generateRandomString(1024)),  // 1 KB
                Arguments.of(generateRandomString(10 * 1024 * 1024)) // 10 MB
            )
        }

        private fun generateRandomString(length: Int): String {
            val random = SecureRandom()
            val sb = StringBuilder(length)
            for (i in 0 until length) {
                val randomAscii = 33 + random.nextInt(94)
                sb.append(randomAscii.toChar())
            }
            return sb.toString()
        }
    }

    @Nested
    @DisplayName("Key Generation Tests")
    internal inner class KeyGenerationTests {
        @Test
        @DisplayName("Test key generation consistency with same password and salt")
        @Throws(
            Exception::class
        )
        fun testKeyGenerationConsistency() {
            // act

            val newKey = generateKeyFromPassword(password, salt)

            // assert
            Assertions.assertArrayEquals(
                key.encoded, newKey.encoded,
                "Keys generated with the same password and salt should be identical"
            )
        }

        @Test
        @DisplayName("Test key generation with different salts produces different keys")
        @Throws(
            Exception::class
        )
        fun testKeyGenerationDifferentSalts() {
            // setup

            val newSalt = generateSalt()

            // act
            val newKey = generateKeyFromPassword(password, newSalt)

            // assert
            Assertions.assertFalse(
                key.encoded.contentEquals(newKey.encoded),
                "Keys generated with the same password but different salts should differ"
            )
        }

        @Test
        @DisplayName("Test key generation with different passwords produces different keys")
        @Throws(
            Exception::class
        )
        fun testKeyGenerationDifferentPasswords() {
            // setup

            val newPassword = "newPassword"

            // act
            val newKey = generateKeyFromPassword(newPassword, salt)

            // assert
            Assertions.assertFalse(
                key.encoded.contentEquals(newKey.encoded),
                "Keys generated with different passwords should differ"
            )
        }
    }

    @Nested
    @DisplayName("Salt Generation Tests")
    internal inner class SaltGenerationTests {
        @Test
        @DisplayName("Test salt is of correct default length")
        fun testDefaultSaltLength() {
            // assert

            Assertions.assertEquals(10, salt.size, "Default salt length should be 10 bytes")
        }

        @ParameterizedTest(name = "Salt length: {0} bytes")
        @ValueSource(ints = [8, 16, 32])
        @DisplayName("Test salt generation with specified lengths")
        fun testSaltGenerationWithSpecifiedLengths(length: Int) {
            // act

            val generatedSalt = generateSalt(length)

            // assert
            Assertions.assertEquals(length, generatedSalt.size, "Salt length should match the specified length")
        }

        @Test
        @DisplayName("Test salt randomness")
        fun testSaltRandomness() {
            // setup

            val numSalts = 1000
            val salts: MutableSet<String> = HashSet()

            // act
            for (i in 0 until numSalts) {
                val salt = generateSalt()
                salts.add(salt.contentToString())
            }

            // assert
            Assertions.assertEquals(numSalts, salts.size, "All generated salts should be unique")
        }
    }

    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
    @Nested
    @DisplayName("Exception Handling Tests")
    internal inner class ExceptionHandlingTests {
        @Test
        @DisplayName("Test decryption with invalid key throws AEADBadTagException")
        @Throws(
            Exception::class
        )
        fun testDecryptionWithInvalidKey() {
            // setup

            val invalidKey = generateKeyFromPassword("invalidPassword", salt)
            val data = "Short text."

            // act
            val encryptedData = encrypt(data.toByteArray(), key)

            // assert
            val exception = Assertions.assertThrows(
                AEADBadTagException::class.java,
                { decrypt(invalidKey, encryptedData.first, encryptedData.second) },
                "Decryption with an invalid key should throw AEADBadTagException"
            )
            Assertions.assertEquals("Tag mismatch", exception.message, "Exception message should indicate tag mismatch")
        }

        @Test
        @DisplayName("Test decryption with invalid IV throws AEADBadTagException")
        @Throws(
            Exception::class
        )
        fun testDecryptionWithInvalidIV() {
            // setup

            val data = "Short text"
            val encryptedData = encrypt(data.toByteArray(), key)
            encryptedData.first[0] = (encryptedData.first[0].toInt() xor 0xFF.toByte().toInt()).toByte()

            // assert
            val exception = Assertions.assertThrows(
                AEADBadTagException::class.java,
                { decrypt(key, encryptedData.first, encryptedData.second) },
                "Decryption with an invalid IV should throw AEADBadTagException"
            )
            Assertions.assertEquals("Tag mismatch", exception.message, "Exception message should indicate tag mismatch")
        }

        @Test
        @DisplayName("Test decryption with corrupted ciphertext throws AEADBadTagException")
        @Throws(
            Exception::class
        )
        fun testDecryptionWithCorruptedCiphertext() {
            // setup

            val data = "Short text"
            val encryptedData = encrypt(data.toByteArray(), key)
            encryptedData.second[0] = (encryptedData.second[0].toInt() xor 0xFF.toByte().toInt()).toByte()

            // assert
            val exception = Assertions.assertThrows(
                AEADBadTagException::class.java,
                { decrypt(key, encryptedData.first, encryptedData.second) },
                "Decryption with corrupted ciphertext should throw AEADBadTagException"
            )
            Assertions.assertEquals("Tag mismatch", exception.message, "Exception message should indicate tag mismatch")
        }

        @ParameterizedTest(name = "Null parameter: {0}")
        @MethodSource("nullParametersProviderForEncryption")
        @DisplayName("Test encryption with null parameters throws NullPointerException")
        fun testEncryptionWithNullParameters(param1: Any?, param2: Any?) {
            // setup

            val data = if (param1 is ByteArray) param1 else null
            val key = if (param2 is SecretKey) param2 else null

            // assert
            Assertions.assertThrows(
                NullPointerException::class.java,
                { encrypt(data!!, key!!) },
                "Encryption with null parameters should throw NullPointerException"
            )
        }

        @ParameterizedTest(name = "Null parameter: {0}")
        @MethodSource("nullParametersProviderForDecryption")
        @DisplayName("Test decryption with null parameters throws NullPointerException")
        fun testDecryptionWithNullParameters(param1: Any?, param2: Any?, param3: Any?) {
            // setup

            val key = if (param1 is SecretKey) param1 else null
            val iv = if (param2 is ByteArray) param2 else null
            val cipherText = if (param3 is ByteArray) param3 else null

            // assert
            Assertions.assertThrows(
                NullPointerException::class.java,
                { decrypt(key!!, iv!!, cipherText!!) },
                "Decryption with null parameters should throw NullPointerException"
            )
        }

        @ParameterizedTest(name = "Null parameter: {0}")
        @MethodSource("nullParametersProviderForKeyGeneration")
        @DisplayName("Test key generation with null parameters throws NullPointerException")
        fun testKeyGenerationWithNullParameters(param1: Any?, param2: Any?) {
            // setup

            val password = if (param1 is String) param1 else null
            val salt = if (param2 is ByteArray) param2 else null

            // assert
            Assertions.assertThrows(
                NullPointerException::class.java,
                { generateKeyFromPassword(password!!, salt!!) },
                "Encryption with null parameters should throw NullPointerException"
            )
        }

        /**
         * Provides null parameters for encryption tests.
         */
        fun nullParametersProviderForEncryption(): Stream<Arguments> {
            return Stream.of(
                Arguments.of(null, "validKey"),
                Arguments.of("validData", null),
                Arguments.of(null, null)
            ).map { args: Arguments ->
                try {
                    val key = generateKeyFromPassword("password", generateSalt())
                    val data = if ("validData" == args.get()[0]) "Short text".toByteArray() else null
                    val keyParam = if ("validKey" == args.get()[1]) key else null
                    return@map Arguments.of(data, keyParam)
                } catch (e: Exception) {
                    throw RuntimeException(e)
                }
            }
        }

        /**
         * Provides null parameters for decryption tests.
         */
        fun nullParametersProviderForDecryption(): Stream<Arguments> {
            return Stream.of(
                Arguments.of(null, "validIV", "validEncryptedData"),
                Arguments.of("validKey", null, "validEncryptedData"),
                Arguments.of("validKey", "validIV", null),
                Arguments.of(null, null, null)
            ).map { args: Arguments ->
                try {
                    val key = generateKeyFromPassword("password", generateSalt())
                    val encryptedData = encrypt("Short text".toByteArray(), key)
                    val keyParam = if ("validKey" == args.get()[0]) key else null
                    val iv = if ("validIV" == args.get()[1]) encryptedData.first else null
                    val cipherText = if ("validEncryptedData" == args.get()[2]) encryptedData.second else null
                    return@map Arguments.of(keyParam, iv, cipherText)
                } catch (e: Exception) {
                    throw RuntimeException(e)
                }
            }
        }

        /**
         * Provides null parameters for key generation tests.
         */
        fun nullParametersProviderForKeyGeneration(): Stream<Arguments> {
            return Stream.of(
                Arguments.of(null, "validSalt"),
                Arguments.of("validPassword", null),
                Arguments.of(null, null)
            ).map { args: Arguments ->
                val passwordParam = if ("validPassword" == args.get()[0]) "password" else null
                val saltParam = if ("validSalt" == args.get()[1]) generateSalt() else null
                Arguments.of(passwordParam, saltParam)
            }
        }
    }
}
